--- nextflow.config	2024-06-26 16:29:53.603738454 +0200
+++ nextflow.config.edited	2024-06-26 16:36:12.282686623 +0200
@@ -32,6 +32,18 @@
     cpus = 32
     memory = 128.GB
   }
+
+  // SP@NC
+  withName: qiime2_phylogeny_diversity {
+    cpus = 8
+    memory = 500.GB
+  }
+
+  // SP@NC
+  withName: html_rep {
+    cpus = 8
+    memory = 128.GB
+  }
 }
 
 
@@ -91,18 +103,21 @@
 
 
 // Generate report
+// file = "report_$params.outdir/report.html"
 report {
   enabled = true
-  file = "report_$params.outdir/report.html"
+  file = "$params.outdir/report/report.html"
+  overwrite = true
 }
 // Timeline
 timeline {
   enabled = true
-  file = "report_$params.outdir/timeline.html"
+  file = "$params.outdir/report/timeline.html"
+  overwrite = true
 }
 // DAG
 dag {
   enabled = true
-  file = "report_$params.outdir/dag.html"
+  file = "$params.outdir/report/dag.html"
   overwrite = true
 }
